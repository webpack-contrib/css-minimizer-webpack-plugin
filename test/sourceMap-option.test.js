import webpack from "webpack";
import MiniCssExtractPlugin from "mini-css-extract-plugin";

import CssMinimizerPlugin from "../src/index";

import {
  getCompiler,
  compile,
  readAssets,
  getErrors,
  getWarnings,
} from "./helpers";

expect.addSnapshotSerializer({
  test: (value) => {
    // For string that are valid JSON
    if (typeof value !== "string") {
      return false;
    }

    try {
      return typeof JSON.parse(value) === "object";
    } catch (e) {
      return false;
    }
  },
  print: (value) => JSON.stringify(JSON.parse(value), null, 4),
});

describe('when applied with "sourceMap" option', () => {
  const baseConfig = {
    devtool: "source-map",
    entry: {
      entry: `${__dirname}/fixtures/sourcemap/foo.scss`,
      entry2: `${__dirname}/fixtures/sourcemap/foo.css`,
    },
    module: {
      rules: [
        {
          test: /.s?css$/i,
          use: [MiniCssExtractPlugin.loader, "css-loader", "sass-loader"],
        },
      ],
    },
    plugins: [
      new MiniCssExtractPlugin({
        filename: "[name].css",
        chunkFilename: "[id].[name].css",
      }),
    ],
  };

  it('should work with the "devtool" option', async () => {
    const compiler = getCompiler(baseConfig);

    new CssMinimizerPlugin().apply(compiler);

    const stats = await compile(compiler);

    expect(readAssets(compiler, stats, /\.css(\.map)?$/)).toMatchSnapshot(
      "assets"
    );
    expect(getErrors(stats)).toMatchSnapshot("errors");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
  });

  it('should work with the "devtool" option and the "parallel" option with "false" value', async () => {
    const compiler = getCompiler(baseConfig);

    new CssMinimizerPlugin({
      parallel: false,
    }).apply(compiler);

    const stats = await compile(compiler);

    expect(readAssets(compiler, stats, /\.css(\.map)?$/)).toMatchSnapshot(
      "assets"
    );
    expect(getErrors(stats)).toMatchSnapshot("errors");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
  });

  it('should work with the "devtool" option and the "parallel" option with "true" value', async () => {
    const compiler = getCompiler(baseConfig);

    new CssMinimizerPlugin({
      parallel: true,
    }).apply(compiler);

    const stats = await compile(compiler);

    expect(readAssets(compiler, stats, /\.css(\.map)?$/)).toMatchSnapshot(
      "assets"
    );
    expect(getErrors(stats)).toMatchSnapshot("errors");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
  });

  it("should work with SourceMapDevToolPlugin plugin)", async () => {
    const config = Object.assign(baseConfig, {
      devtool: false,
      module: {
        rules: [
          {
            test: /.s?css$/i,
            use: [
              MiniCssExtractPlugin.loader,
              { loader: "css-loader", options: { sourceMap: true } },
              { loader: "sass-loader", options: { sourceMap: true } },
            ],
          },
        ],
      },
      plugins: [
        new MiniCssExtractPlugin({
          filename: "dist/[name].css",
          chunkFilename: "dist/[id].[name].css",
        }),
        new webpack.SourceMapDevToolPlugin({
          filename: "sourcemaps/[file].map",
          publicPath: "https://example.com/project/",
          fileContext: "dist",
        }),
      ],
    });

    const compiler = getCompiler(config);

    new CssMinimizerPlugin().apply(compiler);

    const stats = await compile(compiler);

    expect(readAssets(compiler, stats, /\.css(\.map)?$/)).toMatchSnapshot(
      "assets"
    );
    expect(getErrors(stats)).toMatchSnapshot("errors");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
  });

  it("should work and emit warnings on broken sourcemaps", async () => {
    const emitBrokenSourceMapPlugin = new (class EmitBrokenSourceMapPlugin {
      apply(pluginCompiler) {
        pluginCompiler.hooks.compilation.tap(
          { name: this.constructor.name },
          (compilation) => {
            compilation.hooks.processAssets.tap(
              {
                name: this.constructor.name,
                stage: webpack.Compilation.PROCESS_ASSETS_STAGE_ADDITIONAL,
              },
              () => {
                compilation.additionalChunkAssets.push("broken-source-map.css");

                const assetContent = ".bar {color: red};";

                // eslint-disable-next-line no-param-reassign
                compilation.assets["broken-source-map.css"] = {
                  size() {
                    return assetContent.length;
                  },
                  source() {
                    return assetContent;
                  },
                  sourceAndMap() {
                    return {
                      source: this.source(),
                      map: {
                        sources: [],
                        names: [],
                        mappings: "AAAA,KAAK,iBAAiB,KAAK,UAAU,OAAO",
                        file: "x",
                        sourcesContent: [],
                      },
                    };
                  },
                };
              }
            );
          }
        );
      }
    })();

    const config = Object.assign(baseConfig, {
      devtool: "source-map",
      entry: {
        entry2: `${__dirname}/fixtures/sourcemap/foo.css`,
      },
      plugins: [
        emitBrokenSourceMapPlugin,
        new MiniCssExtractPlugin({
          filename: "dist/[name].css",
          chunkFilename: "dist/[id].[name].css",
        }),
      ],
    });

    const compiler = getCompiler(config);

    new CssMinimizerPlugin().apply(compiler);

    const stats = await compile(compiler);

    expect(readAssets(compiler, stats, /\.css(\.map)?$/)).toMatchSnapshot(
      "assets"
    );
    expect(getErrors(stats)).toMatchSnapshot("errors");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(stats.compilation.warnings[0].toString()).toMatch(
      "contains invalid source map"
    );
  });

  it("should work and emit warning on valid sourcemap and minimizer error", async () => {
    const emitBrokenSourceMapPlugin = new (class EmitBrokenSourceMapPlugin {
      apply(pluginCompiler) {
        pluginCompiler.hooks.compilation.tap(
          { name: this.constructor.name },
          (compilation) => {
            compilation.hooks.processAssets.tap(
              {
                name: this.constructor.name,
                stage: webpack.Compilation.PROCESS_ASSETS_STAGE_ADDITIONAL,
              },
              () => {
                compilation.additionalChunkAssets.push("broken-source-map.css");

                const assetContent = ".bar {color: red};";

                // eslint-disable-next-line no-param-reassign
                compilation.assets["broken-source-map.css"] = {
                  size() {
                    return assetContent.length;
                  },
                  source() {
                    return assetContent;
                  },
                  sourceAndMap() {
                    return {
                      source: this.source(),
                      map: {
                        version: 3,
                        sources: ["test", "test2"],
                        names: [],
                        mappings: "AAAA,KAAK,iBAAiB,KAAK,UAAU,OAAO",
                        file: "x",
                        sourcesContent: [],
                      },
                    };
                  },
                };
              }
            );
          }
        );
      }
    })();

    const config = Object.assign(baseConfig, {
      devtool: "source-map",
      entry: {
        entry2: `${__dirname}/fixtures/sourcemap/foo.css`,
      },
      plugins: [
        emitBrokenSourceMapPlugin,
        new MiniCssExtractPlugin({
          filename: "dist/[name].css",
          chunkFilename: "dist/[id].[name].css",
        }),
      ],
    });

    const compiler = getCompiler(config);

    new CssMinimizerPlugin({
      minify: (data) => {
        // eslint-disable-next-line global-require
        const postcss = require("postcss");

        const plugin = () => {
          let erroredDecl;

          return {
            postcssPlugin: "error-plugin",
            Declaration(decl) {
              erroredDecl = decl;
            },
            OnceExit() {
              throw erroredDecl.error("Postcss error");
            },
          };
        };

        plugin.postcss = true;

        const [[filename, input]] = Object.entries(data);

        return postcss([plugin])
          .process(input, { from: filename, to: filename })
          .then((result) => result);
      },
    }).apply(compiler);

    const stats = await compile(compiler);

    expect(getErrors(stats)).toMatchSnapshot("errors");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
  });

  it("should work and do not contain sourcemap link in minified source", async () => {
    const compiler = getCompiler({
      devtool: false,
      entry: {
        foo: `${__dirname}/fixtures/foo.css`,
      },
    });

    new CssMinimizerPlugin().apply(compiler);

    const stats = await compile(compiler);
    const assets = readAssets(compiler, stats, /\.css$/);
    const [[, input]] = Object.entries(assets);

    expect(/sourceMappingURL/i.test(input)).toBe(false);
  });
});
