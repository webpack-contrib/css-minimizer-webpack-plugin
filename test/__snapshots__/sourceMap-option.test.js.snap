// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when applied with "sourceMap" option should work and emit warning on valid sourcemap and minimizer error: errors 1`] = `
Array [
  "Error: broken-source-map.css from Css Minimizer plugin
error-plugin: /broken-source-map.css:1:7: Postcss error [test:1,5][broken-source-map.css:1,7]",
  "Error: dist/entry2.css from Css Minimizer plugin
error-plugin: /dist/entry2.css:10:3: Postcss error [webpack://./sourcemap/foo.css:8,2][dist/entry2.css:10,3]",
]
`;

exports[`when applied with "sourceMap" option should work and emit warning on valid sourcemap and minimizer error: warnings 1`] = `Array []`;

exports[`when applied with "sourceMap" option should work and emit warnings on broken sourcemaps: assets 1`] = `
Object {
  "broken-source-map.css": ".bar{color:red};",
  "dist/entry2.css": "body{color:red;font-weight:700}body a{text-align:center}
/*# sourceMappingURL=entry2.css.map*/",
  "dist/entry2.css.map": {
    "version": 3,
    "file": "dist/entry2.css",
    "mappings": "AAAA,KAKE,SAAU,CAJV,eACF,CAMA,OACE,iBACF",
    "sources": [
        "webpack:///entry2.css"
    ],
    "sourcesContent": [
        "body {\\n  font-weight: bold;\\n}\\n\\nbody {\\n  color: red;\\n}\\n\\nbody a {\\n  text-align: center;\\n}\\n"
    ],
    "names": [],
    "sourceRoot": ""
},
}
`;

exports[`when applied with "sourceMap" option should work and emit warnings on broken sourcemaps: errors 1`] = `Array []`;

exports[`when applied with "sourceMap" option should work and emit warnings on broken sourcemaps: warnings 1`] = `
Array [
  "Error: broken-source-map.css contains invalid source map",
]
`;

exports[`when applied with "sourceMap" option should work with SourceMapDevToolPlugin plugin): assets 1`] = `
Object {
  "dist/entry.css": "body{color:red;font-weight:700}body a{text-align:center}
/*# sourceMappingURL=https://example.com/project/sourcemaps/entry.css.map*/",
  "dist/entry2.css": "body{color:red;font-weight:700}body a{text-align:center}
/*# sourceMappingURL=https://example.com/project/sourcemaps/entry2.css.map*/",
  "sourcemaps/entry.css.map": {
    "version": 3,
    "file": "dist/entry.css",
    "mappings": "AAAA,KAKE,SAAU,CAJV,eACF,CAKA,OACE,iBACF",
    "sources": [
        "webpack:///entry.css"
    ],
    "sourcesContent": [
        "body {\\n  font-weight: bold;\\n}\\n\\nbody {\\n  color: red;\\n}\\nbody a {\\n  text-align: center;\\n}\\n"
    ],
    "names": [],
    "sourceRoot": ""
},
  "sourcemaps/entry2.css.map": {
    "version": 3,
    "file": "dist/entry2.css",
    "mappings": "AAAA,KAKE,SAAU,CAJV,eACF,CAMA,OACE,iBACF",
    "sources": [
        "webpack:///entry2.css"
    ],
    "sourcesContent": [
        "body {\\n  font-weight: bold;\\n}\\n\\nbody {\\n  color: red;\\n}\\n\\nbody a {\\n  text-align: center;\\n}\\n"
    ],
    "names": [],
    "sourceRoot": ""
},
}
`;

exports[`when applied with "sourceMap" option should work with SourceMapDevToolPlugin plugin): errors 1`] = `Array []`;

exports[`when applied with "sourceMap" option should work with SourceMapDevToolPlugin plugin): warnings 1`] = `Array []`;

exports[`when applied with "sourceMap" option should work with the "devtool" option and the "parallel" option with "false" value: assets 1`] = `
Object {
  "entry.css": "body{color:red;font-weight:700}body a{text-align:center}
/*# sourceMappingURL=entry.css.map*/",
  "entry.css.map": {
    "version": 3,
    "file": "entry.css",
    "mappings": "AAAA,KCGE,SAAU,CDFV,eCCF,CAEE,OACE,iBAIJ",
    "sources": [
        "webpack:///./sourcemap/bar.scss",
        "webpack:///./sourcemap/foo.scss"
    ],
    "sourcesContent": [
        "body {\\n  font-weight: bold;\\n}",
        "@import 'bar';\\n\\nbody {\\n  color: red;\\n  a {\\n    text-align: center;\\n  }\\n}"
    ],
    "names": [],
    "sourceRoot": ""
},
  "entry2.css": "body{color:red;font-weight:700}body a{text-align:center}
/*# sourceMappingURL=entry2.css.map*/",
  "entry2.css.map": {
    "version": 3,
    "file": "entry2.css",
    "mappings": "AAAA,KCGE,SAAU,CDFV,eCCF,CAIA,OACE,iBAGF",
    "sources": [
        "webpack:///./sourcemap/bar.scss",
        "webpack:///./sourcemap/foo.css"
    ],
    "sourcesContent": [
        "body {\\n  font-weight: bold;\\n}",
        "@import 'bar';\\n\\nbody {\\n  color: red;\\n}\\n\\nbody a {\\n  text-align: center;\\n}"
    ],
    "names": [],
    "sourceRoot": ""
},
}
`;

exports[`when applied with "sourceMap" option should work with the "devtool" option and the "parallel" option with "false" value: errors 1`] = `Array []`;

exports[`when applied with "sourceMap" option should work with the "devtool" option and the "parallel" option with "false" value: warnings 1`] = `Array []`;

exports[`when applied with "sourceMap" option should work with the "devtool" option and the "parallel" option with "true" value: assets 1`] = `
Object {
  "entry.css": "body{color:red;font-weight:700}body a{text-align:center}
/*# sourceMappingURL=entry.css.map*/",
  "entry.css.map": {
    "version": 3,
    "file": "entry.css",
    "mappings": "AAAA,KCGE,SAAU,CDFV,eCCF,CAEE,OACE,iBAIJ",
    "sources": [
        "webpack:///./sourcemap/bar.scss",
        "webpack:///./sourcemap/foo.scss"
    ],
    "sourcesContent": [
        "body {\\n  font-weight: bold;\\n}",
        "@import 'bar';\\n\\nbody {\\n  color: red;\\n  a {\\n    text-align: center;\\n  }\\n}"
    ],
    "names": [],
    "sourceRoot": ""
},
  "entry2.css": "body{color:red;font-weight:700}body a{text-align:center}
/*# sourceMappingURL=entry2.css.map*/",
  "entry2.css.map": {
    "version": 3,
    "file": "entry2.css",
    "mappings": "AAAA,KCGE,SAAU,CDFV,eCCF,CAIA,OACE,iBAGF",
    "sources": [
        "webpack:///./sourcemap/bar.scss",
        "webpack:///./sourcemap/foo.css"
    ],
    "sourcesContent": [
        "body {\\n  font-weight: bold;\\n}",
        "@import 'bar';\\n\\nbody {\\n  color: red;\\n}\\n\\nbody a {\\n  text-align: center;\\n}"
    ],
    "names": [],
    "sourceRoot": ""
},
}
`;

exports[`when applied with "sourceMap" option should work with the "devtool" option and the "parallel" option with "true" value: errors 1`] = `Array []`;

exports[`when applied with "sourceMap" option should work with the "devtool" option and the "parallel" option with "true" value: warnings 1`] = `Array []`;

exports[`when applied with "sourceMap" option should work with the "devtool" option: assets 1`] = `
Object {
  "entry.css": "body{color:red;font-weight:700}body a{text-align:center}
/*# sourceMappingURL=entry.css.map*/",
  "entry.css.map": {
    "version": 3,
    "file": "entry.css",
    "mappings": "AAAA,KCGE,SAAU,CDFV,eCCF,CAEE,OACE,iBAIJ",
    "sources": [
        "webpack:///./sourcemap/bar.scss",
        "webpack:///./sourcemap/foo.scss"
    ],
    "sourcesContent": [
        "body {\\n  font-weight: bold;\\n}",
        "@import 'bar';\\n\\nbody {\\n  color: red;\\n  a {\\n    text-align: center;\\n  }\\n}"
    ],
    "names": [],
    "sourceRoot": ""
},
  "entry2.css": "body{color:red;font-weight:700}body a{text-align:center}
/*# sourceMappingURL=entry2.css.map*/",
  "entry2.css.map": {
    "version": 3,
    "file": "entry2.css",
    "mappings": "AAAA,KCGE,SAAU,CDFV,eCCF,CAIA,OACE,iBAGF",
    "sources": [
        "webpack:///./sourcemap/bar.scss",
        "webpack:///./sourcemap/foo.css"
    ],
    "sourcesContent": [
        "body {\\n  font-weight: bold;\\n}",
        "@import 'bar';\\n\\nbody {\\n  color: red;\\n}\\n\\nbody a {\\n  text-align: center;\\n}"
    ],
    "names": [],
    "sourceRoot": ""
},
}
`;

exports[`when applied with "sourceMap" option should work with the "devtool" option: errors 1`] = `Array []`;

exports[`when applied with "sourceMap" option should work with the "devtool" option: warnings 1`] = `Array []`;
